File: apps/ios/Shared/Views/Chat/Group/ChannelRelaysView.swift
//
//  ChannelRelaysView.swift
//  SimpleX (iOS)
//
//  Created by spaced4ndy on 20.02.2026.
//  Copyright © 2026 SimpleX Chat. All rights reserved.
//

import SwiftUI
import SimpleXChat

struct ChannelRelaysView: View {
    @ObservedObject var chat: Chat
    var groupInfo: GroupInfo
    @EnvironmentObject var chatModel: ChatModel
    @EnvironmentObject var theme: AppTheme
    @ObservedObject private var channelRelaysModel = ChannelRelaysModel.shared
    @State private var showAddRelay = false

    private var groupRelays: [GroupRelay] {
        channelRelaysModel.groupId == groupInfo.groupId ? channelRelaysModel.groupRelays : []
    }

    var body: some View {
        List {
            relaysList()
            // TODO [relays] re-enable when relay management ships
            // if groupInfo.isOwner {
            //     Section {
            //         Button {
            //             showAddRelay = true
            //         } label: {
            //             Label("Add relay", systemImage: "plus")
            //         }
            //     }
            // }
        }
        // TODO [relays] re-enable when relay management ships
        // .sheet(isPresented: $showAddRelay) {
        //     // Backend gate (APIAddGroupRelays) rejects any chatRelayId already in group_relays
        //     // regardless of relayStatus, so all current rows must be excluded from the add list.
        //     let existingRelayIds = Set(groupRelays.compactMap { $0.userChatRelay.chatRelayId })
        //     AddGroupRelayView(groupInfo: groupInfo, existingRelayIds: existingRelayIds) {
        //         Task { await chatModel.loadGroupMembers(groupInfo) }
        //     }
        // }
        .onAppear {
            Task {
                await chatModel.loadGroupMembers(groupInfo)
                if groupInfo.isOwner {
                    let relays = await apiGetGroupRelays(groupInfo.groupId)
                    await MainActor.run {
                        ChannelRelaysModel.shared.set(groupId: groupInfo.groupId, groupRelays: relays)
                    }
                }
            }
        }
    }

    @ViewBuilder private func relaysList() -> some View {
        let relayMembers = chatModel.groupMembers.filter { $0.wrapped.memberRole == .relay && $0.wrapped.memberStatus != .memRemoved && $0.wrapped.memberStatus != .memGroupDeleted }
        if relayMembers.isEmpty {
            Section {
                Text("No chat relays")
                    .foregroundColor(theme.colors.secondary)
            }
        } else {
            Section {
                ForEach(relayMembers) { member in
                    let link = NavigationLink {
                        GroupMemberInfoView(
                            groupInfo: groupInfo,
                            chat: chat,
                            groupMember: member,
                            scrollToItemId: Binding.constant(nil),
                            groupRelay: groupRelays.first(where: { $0.groupMemberId == member.wrapped.groupMemberId })
                        )
                        .navigationBarHidden(false)
                    } label: {
                        let statusText = groupInfo.isOwner
                            ? ownerRelayStatusText(member.wrapped)
                            : subscriberRelayStatusText(member.wrapped)
                        relayMemberRow(member.wrapped, statusText: statusText)
                    }
                    // TODO [relays] re-enable when relay management ships
                    // if groupInfo.isOwner && member.wrapped.canBeRemoved(groupInfo: groupInfo) {
                    //     link.swipeActions(edge: .trailing) {
                    //         Button {
                    //             showRemoveMemberAlert(groupInfo, member.wrapped)
                    //         } label: {
                    //             Label("Remove relay", systemImage: "trash")
                    //         }
                    //         .tint(.red)
                    //     }
                    // } else {
                    //     link
                    // }
                    link
                }
            } footer: {
                Text("Chat relays forward messages to channel subscribers.")
            }
        }
    }

    private func subscriberRelayStatusText(_ member: GroupMember) -> LocalizedStringKey {
        if member.activeConn?.connDisabled ?? false {
            "disabled"
        } else if member.activeConn?.connInactive ?? false {
            "inactive"
        } else {
            relayConnStatus(member).text
        }
    }

    private func ownerRelayStatusText(_ member: GroupMember) -> LocalizedStringKey {
        let relayStatus = groupRelays.first(where: { $0.groupMemberId == member.groupMemberId })?.relayStatus
        return if relayStatus == .rejected {
            "rejected"
        } else if [.memLeft, .memRemoved, .memGroupDeleted].contains(member.memberStatus) {
            relayConnStatus(member).text
        } else if case .failed = member.activeConn?.connStatus {
            "failed"
        } else if member.activeConn?.connDisabled ?? false {
            "disabled"
        } else if member.activeConn?.connInactive ?? false {
            "inactive"
        } else {
            relayStatus?.text ?? relayConnStatus(member).text
        }
    }

    private func relayMemberRow(_ member: GroupMember, statusText: LocalizedStringKey) -> some View {
        HStack {
            MemberProfileImage(member, size: 38)
                .padding(.trailing, 2)
            VStack(alignment: .leading) {
                Text(member.chatViewName)
                    .foregroundColor(theme.colors.onBackground)
                    .lineLimit(1)
                Text(statusText)
                    .lineLimit(1)
                    .font(.caption)
                    .foregroundColor(theme.colors.secondary)
            }
            Spacer()
        }
    }
}

func relayConnStatus(_ member: GroupMember) -> (text: LocalizedStringKey, color: Color) {
    switch member.memberStatus {
    case .memLeft: ("removed by operator", .red)
    case .memRemoved, .memGroupDeleted: (member.memberStatus.text, .red)
    default:
        switch member.activeConn?.connStatus {
        case .ready: ("connected", .green)
        case .deleted: ("deleted", .red)
        case .failed: ("failed", .red)
        default: ("connecting", .yellow)
        }
    }
}

func hostFromRelayLink(_ link: String) -> String {
    if let ft = parseSimpleXMarkdown(link) {
        for f in ft {
            if case let .simplexLink(_, _, _, smpHosts) = f.format,
               let host = smpHosts.first {
                return host
            }
        }
    }
    return link
}

#Preview {
    ChannelRelaysView(chat: Chat.sampleData, groupInfo: GroupInfo.sampleData)
}

File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/chat/group/ChannelRelaysView.kt
package chat.simplex.common.views.chat.group

import SectionBottomSpacer
import SectionItemView
import SectionTextFooter
import SectionView
import androidx.compose.foundation.layout.*
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import chat.simplex.common.model.*
import chat.simplex.common.platform.*
import chat.simplex.common.ui.theme.*
import chat.simplex.common.views.chatlist.setGroupMembers
import chat.simplex.common.views.helpers.*
import chat.simplex.res.MR

@Composable
fun ChannelRelaysView(
  rhId: Long?,
  groupInfo: GroupInfo,
  chatModel: ChatModel,
  close: () -> Unit,
  showMemberInfo: (GroupMember, GroupRelay?) -> Unit
) {
  BackHandler(onBack = close)
  var groupRelays by remember { mutableStateOf<List<GroupRelay>>(emptyList()) }

  LaunchedEffect(Unit) {
    setGroupMembers(rhId, groupInfo, chatModel)
    if (groupInfo.isOwner) {
      groupRelays = chatModel.controller.apiGetGroupRelays(groupInfo.groupId)
    }
  }

  ChannelRelaysLayout(
    groupInfo = groupInfo,
    chatModel = chatModel,
    groupRelays = groupRelays,
    showMemberInfo = showMemberInfo
  )
}

@Composable
private fun ChannelRelaysLayout(
  groupInfo: GroupInfo,
  chatModel: ChatModel,
  groupRelays: List<GroupRelay>,
  showMemberInfo: (GroupMember, GroupRelay?) -> Unit
) {
  val relayMembers = remember { chatModel.groupMembers }.value
    .filter { it.memberRole == GroupMemberRole.Relay }

  ColumnWithScrollBar {
    AppBarTitle(generalGetString(MR.strings.channel_relays_title))

    if (relayMembers.isEmpty()) {
      SectionView {
        SectionItemView(padding = PaddingValues(horizontal = DEFAULT_PADDING)) {
          Text(
            generalGetString(MR.strings.no_chat_relays),
            color = MaterialTheme.colors.secondary
          )
        }
      }
    } else {
      SectionView {
        relayMembers.forEachIndexed { index, member ->
          if (index > 0) {
            Divider()
          }
          SectionItemView(
            click = { showMemberInfo(member, groupRelays.firstOrNull { it.groupMemberId == member.groupMemberId }) },
            minHeight = 54.dp,
            padding = PaddingValues(horizontal = DEFAULT_PADDING)
          ) {
            val statusText = if (groupInfo.isOwner) {
              ownerRelayStatusText(member, groupRelays)
            } else {
              subscriberRelayStatusText(member)
            }
            RelayMemberRow(member, statusText)
          }
        }
      }
      SectionTextFooter(generalGetString(MR.strings.chat_relays_forward_messages))
    }
    SectionBottomSpacer()
  }
}

@Composable
private fun RelayMemberRow(member: GroupMember, statusText: String) {
  Row(
    Modifier.fillMaxWidth(),
    verticalAlignment = Alignment.CenterVertically,
    horizontalArrangement = Arrangement.spacedBy(4.dp)
  ) {
    MemberProfileImage(size = 38.dp, member)
    Spacer(Modifier.width(2.dp))
    Column(Modifier.weight(1f)) {
      Text(
        member.chatViewName,
        maxLines = 1,
        overflow = TextOverflow.Ellipsis,
        color = MaterialTheme.colors.onBackground
      )
      Text(
        statusText,
        maxLines = 1,
        fontSize = 12.sp,
        color = MaterialTheme.colors.secondary
      )
    }
  }
}

private fun subscriberRelayStatusText(member: GroupMember): String {
  return if (member.activeConn?.connDisabled == true) {
    generalGetString(MR.strings.member_info_member_disabled)
  } else if (member.activeConn?.connInactive == true) {
    generalGetString(MR.strings.member_info_member_inactive)
  } else {
    relayConnStatus(member).first
  }
}

private fun ownerRelayStatusText(member: GroupMember, groupRelays: List<GroupRelay>): String {
  return if (member.activeConn?.connStatus is ConnStatus.Failed) {
    generalGetString(MR.strings.relay_conn_status_failed)
  } else if (member.activeConn?.connDisabled == true) {
    generalGetString(MR.strings.member_info_member_disabled)
  } else if (member.activeConn?.connInactive == true) {
    generalGetString(MR.strings.member_info_member_inactive)
  } else {
    groupRelays.firstOrNull { it.groupMemberId == member.groupMemberId }?.relayStatus?.text
      ?: relayConnStatus(member).first
  }
}

fun relayConnStatus(member: GroupMember): Pair<String, Color> {
  return when (member.activeConn?.connStatus) {
    is ConnStatus.Ready -> generalGetString(MR.strings.relay_conn_status_connected) to Color.Green
    is ConnStatus.Deleted -> generalGetString(MR.strings.relay_conn_status_deleted) to Color.Red
    is ConnStatus.Failed -> generalGetString(MR.strings.relay_conn_status_failed) to Color.Red
    else -> generalGetString(MR.strings.relay_conn_status_connecting) to WarningOrange
  }
}

fun hostFromRelayLink(link: String): String {
  val ft = parseToMarkdown(link)
  if (ft != null) {
    for (f in ft) {
      val format = f.format
      if (format is Format.SimplexLink) {
        val host = format.smpHosts.firstOrNull()
        if (host != null) return host
      }
    }
  }
  return link
}
